README

                     Tiny Applications Collection

Tiny Applications Collection is a set of small programs, shell scripts
and  other things  which  may be  useful  in everyday  life of  system
administrators, users  and programmers.   They can be  also used  as a
base for a more specialised application as they are all Open Source.

If you have written utility which may be useful for others and want to
share it please email me  at mina86/AT/mina86.com, I'll be glad to add
your utility to the collection.

At the moment the collection includes:

 * add - adds specified text to each line from stdin;

 * ai-pitr.pl - it's  a joke really.  A 6 line long  AI which might be
   written    by   Pitr    from   UserFriendly.     For    more   see:
   http://ars.userfriendly.org/cartoons/?id=19990602 ;

 * ai-sid.pl - it's a joke as well.   An 8 line long AI which might by
   written    by   Sid    from   UserFriendly.     Fore    more   see:
   http://ars.userfriendly.org/cartoons/?id=20001014 ;

 * arpping - tool which pings device with given IP using ARP protocol.
   It may be used without IP address assigned to our machine,
   ie. priori to calling DHCP client or such.

 * changelog.pl  -   application  providing  global   change  log  for
   administrators to  fill and  users to read.   For instance,  can be
   used in /etc/profile to show all news;

 * cdiff (C and sed versions) -  adds ANSI codes to the output of diff
   making it a bit more readable.

 * check.sh  -  performs  a  defined checking  (whether  www.googl.com
   responds to pings by default) and runs specified command if failed.
   May be used to automatically  shut down PC when connection is lost,
   try to reconnect, monitor a server machine and send message to it's
   admin if it's down, keep given application running and many more;

 * checkmail -  checks whether mboxes  given as arguments  contain any
   urnead mail;

 * cpuload.sh - tiny  script showing the CPU load.   May be useful for
   people  interested in  how  it  is calculated  (see  also load  and
   traf.sh);

 * cutcom - removes C++ comments  from stdin and outputs the result to
   stdout;

 * drun  - it's  really  'null' with  aditional  ability to  daemonize
   (hence  the  name).   Helpful  when  running  X  applications  from
   a terminal.

 * errno  - a  tool for  converting  between errno  error numbers  and
   symbolic names which also prints error message.

 * extractlinks.pl  - extracts links  from a  HTML page.   Useful, for
   example, when downloading an Internet gallery;

 * FvwmTransFocus  - an  FVWM  module which  changes windows'  opacity
   depending  on focus.   When  window looses  focus  it becomes  more
   transparent which looks "neet";

 * genpass.sh - generates a random 64-character long password;

 * get_mks_vir_bases -  downlaods MKS-Vir virus  database.  Simplified
   version  of mksupdate  shell script  (from mks_vir  antivirus).  It
   does only download up-to-date virus bases and nothing more;

 * getlyrics.pl - downloads  lyrics of a given song  from the Internet
   (http://www.lyrc.com.ar/ actually);

 * gz2bz - a shell  script which allows converting between compression
   formats.  It  started as gzip  -> bzip2 converter but  now supports
   conversion  between compress,  gzip, bzip2,  xz, lzma  and anything
   else you specify.

 * infinite-logger - loggs every logged message including the messages
   logget by itself :]

 * inplace -  a shell script which  runs given commmand  for all given
   files and replace their content with whatever the command returns;

 * installkernel  - pretty  complex  script for  installing the  Linux
   kernel.   It  makes  a  lot  of aspects  of  upgrading  the  kernel
   automatic.  The first versions required bash to run but the now, it
   is believed  to run on any  POSIX compatible platform so  it can be
   run even on small embeded systems;

 * ivona.sh - uses IVONA speach synthetiser to say given text.  Script
   uses  demo  available  at  http://www.ivo.pl/  so  it  requires  an
   Interent conection to work unless  it has all required MP3s cached.
   IVONA is  a Polish synthesiser so  the script is  rather useful for
   Poles though it may be used as a base for some English synthesiser.

 * lesspipe -  an input processor  for less pager  which automatically
   decompresses  compressed  files, list  contents  of archives,  list
   contents of a  directory, download file from the  Internet and many
   more.

 * load -  shows CPU  load, memory and  net traffic  statistics.  It's
   mostly  ment for  people who  are  curious how  to callculate  such
   statistics and use  some code from the file  in their programs (see
   also cpuload.sh and traf.sh);

 * malloc -  a small program  which allocates memory.   Sounds stupid?
   Mayby, but in theory it can help freeing RAM and therefore speeding
   up the machine;

 * mountiso  - a simple  mount/umount wrapper  for mounting/unmounting
   ISO images  or any other files  using loop option.  You  have to be
   root to use '-o loop' so if you want to mount an ISO image you have
   to login as  root.  This utility should be  SUIDed and then anyoune
   can use it to mount ISO images.  By default it is not included when
   installing tinyapps using 'make  install' or creating package - you
   need to install it by yourself.

 * moz2elinks.pl  -  converts  bookmarks  from  Mozilla's  to  ELinks'
   (http://elinks.or.cz) format;

 * mp3rip - a  command line utility for ripping  tracks from CD Audio.
   It uses  cdparanoia to rip  the tracks and  lame to encode  them to
   MP3s.

 * mpd-show - tool meant to  display song currently played in MPD.  It
   displays it  in a terminal  in one line with  background indicating
   the progress.  It  requires libmpdclient.h and libmpdclient.c files
   to compile;

 * mpd-state  - a  program based  on Avuton  Olrich's  state-utils for
   saving and restoring MPD  (Music Player Daemon) state.  It requires
   libmpdclient.h and libmpdclient.c files to compile;

 * mpd-state-wrapper.sh  -  a  small  shell  script  making  mpd-state
   behaving like Avuton Olrich's state-utils.

 * null  - a C  program (null.c)  and a  shell script  (null.sh) which
   discard  standard input.   It is  meant for  people who  thnik that
   '>/dev/null' is too long to type  every time. :) It may also be run
   with arguments  and the given program  will be run  and both stdout
   and stderr will be redirected to /dev/null;

 * one-time-pass.sh -  another joke.  It  works like passwd  with this
   difference that it automatically  generates a random password which
   is  not revealed to  the user,  therefore he  doesn't know  his own
   password;

 * pingrange.pl - pings a range of addresses;

 * quotes.txt - a collection of quotations (some are in Polish);

 * rand-files.pl - chooses a random set of files, where each file may
   have its own rank or probability, read from standard input such
   that their size is no greater then given limit.  Useful for
   choosing random set of songs to copy to our portable player.

 * rot13 - performs an ROT13 'encryption' on stdin;

 * settitle -  sets the title of  xterm, rxvt, Eterm,  aterm and other
   xterm-compatible terminals;

 * show  - shows  colourised  (if  ccza is  available)  log file  with
   possible option of searching for  a pattern.  Small but handy - you
   don't  have to write  `less /var/log/syslog`  anymore and  get ugly
   black and white log;

 * SlackBuild.gen  -  generates a  SlackBuild  template.   It was  not
   really debugged yet but I believe that it'll work quit good;

 * splitlines.sh -  Splits input, line-by-line, into  several files so
   that the 1st line goes to the 1st file, the 2nd to the 2nd file and
   so on till it wraps and n+1st line goes to 1st file.  I use it when
   I want  to download a lot of  small files using a  few instances of
   wget - it's enough to split file  with urls to 2 or 3 files and run
   2/3 instances of wget in the background.

 * the-book-of-mozilla.txt - a collection  of quotations from The Book
   of Mozilla (most are unofficial);

 * timer - a text-mode timer counting with accuracy to 0.01 s;

 * tpwd -  truncates PWD  to specified number  of characters.   May be
   used in PS1  to display only the  end of the PWD instead  of a full
   60-character long PWD which doesn't really look nice;

 * traf.sh - tiny script  showing traffic statistics.  Many things are
   hardcoded  in  the   source  but  it  may  be   useful  for  future
   modifications  or to  learn  how traffic  is  calculated (see  also
   cpuload.sh and load);

 * tuptime  - my  personal favourite  :)  Displays total  (sum of  all
   uptimes)  and biggest  uptime.   Might be  used  as replacement  to
   standard uptime utility;

 * tv - Perl script showing  TV programme.  It gets data from tv.o2.pl
   so it's usable for Poles mostly  but in the future it is planned to
   provide plugins allowing to download data from other sites.

 * virtman.sh -  adds virtual hosts to Apache  configuration for given
   user.  May be added to  some kind of adduser script for automagical
   addition of virtual hosts.

 * xcolor2rgb  -  reads from  argument  list  (or  from stdin,  if  no
   arguments are given) and prints appropriate RGB value to stdout.

 * xgetclass - prints window's  application name and class as returned
   by XGetClassHint.   May be used in  some kind of  shell scripts but
   honestly I just  wrote it to test XGetClassHint  function and since
   I did it I decided to include it in tinyapps. :)

More  documentation about  each  utility is  usually  provided in  the
source code.


ChangeLog
2010-06-14  Michal Nazarewicz  <mina86/AT/mina86.com>

	* errno: New file

2010-03-21  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mpd-show.c: + fixed progress stopping at the end of the title bug
	+ fixed errors being displayed incorrectly

2010-03-17  Michal Nazarewicz  <mina86/AT/mina86.com>

	* cdiff.c: + better handling of recursive diff
	+ code refactored

2010-03-14  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mpd-show.c: + added iconv support

2010-03-12  Michal Nazarewicz  <mina86/AT/mina86.com>

	* ChangeLog: + spelling corrected
	+ removed references to TODO file

	* gz2bz: + added support for other archivers
	* Makefile, README: + updated gz2bz

2010-03-08  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mpd-show.c: + fixed bug with song formatting
	+ added %Y% pseudo tag which works as if it was not empty but is empty
	+ reverted to old non-highlighted colour

	* mpd-show.c: + small fixes in UTF-8 and wchar_t handling

2010-03-02  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mpd-show.c: + SIGWINCH signal (thus resizing) handling added
	+ some UTF-8 handling added

2010-01-03  Michal Nazarewicz  <mina86/AT/mina86.com>

	* getlyrics.pl: + incorporated lyrics retrieving code by Minio

	* splitlines.sh: + implemented *much* faster method of splitting

2009-12-12  Michal Nazarewicz  <mina86/AT/mina86.com>

	* installkernel: + --*config is now synonym of --config=*config

2009-12-12  Michal Nazarewicz  <mina86/AT/mina86.com>

	* Makefile: + fixed 'package' target

2009-11-25  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mkpkg: + updated to create .txz packages

2009-08-31  Michal Nazarewicz  <mina86/AT/mina86.com>

	* installkernel: + fixed --jobs argument handling

2009-08-30  Michal Nazarewicz  <mina86/AT/mina86.com>

	* installkernel: + added --jobs option
	+ fixed incorrect .version being read when --out used
	+ fixed bug with --su when SHELL pointed to zsh

2009-06-24  Michal Nazarewicz  <mina86/AT/mina86.com>

	* lesspipe: + fixed bug where compressed files were not decompressed

2009-06-14  Michal Nazarewicz  <mina86/AT/mina86.com>

	* gz2bz: + added possibility to save uncompressed files in
	specified directory (think /tmp)
	+ changed meaning of -K (is now --keep-gz, was --keep-both)

2009-05-22  Michal Nazarewicz  <mina86/AT/mina86.com>

	* lesspipe: + added lzma and xz support

	* installkernel, installkernel.8: + added -o parameter

2009-05-09  Michal Nazarewicz  <mina86/AT/mina86.com>

	* quotes.txt: + converted to UTF-8

	* mpd-show.c: + uses fwrite() instead of write(), etc.

2009-04-14  Michal Nazarewicz  <mina86/AT/mina86.com>

	* tpwd: + added --zsh option

	* genpass.sh: + increased portability
	+ added sha1sum as an alternative if md5sum is missing

	* rand-files.pl: + partially rewritten
	+ added ranking
	+ removed "find" functionality (TODO)

	* installkernel: + fixed bug where "su" was interpreted as "-su"
	when "--" was in effect

2009-04-11  Michal Nazarewicz  <mina86/AT/mina86.com>

	cdiff.c: + added isatty() checking (no colours if stdout is not a TTY)

2008-11-08  Michal Nazarewicz  <mina86/AT/mina86.com>

	* quotes.txt: + its "Oscar Wilde" not "Oscar Wide", dunno how this
	typo show up in the first place
	+ a few new quotes

	* LICENSE.GPL3: New file

	+ changed license (from GPLv2 to GPLv3) of some of the files
	+ updated comments of many files (added "This is part of TAC")

	* LICENSE.AFL: + updated to AFL 3.0

2008-11-01  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mountiso.c: + added large file support

	* Makefile: + mountiso mode changed to 4750 and ownership to
	root:root (ie. fewer users can use it by default now)
	+ removed cdiff.sed, get_mks_vir_bases and ivona.sh from default
	install target

2008-07-25  Michal Nazarewicz  <mina86/AT/mina86.com>

	* rand-files.pl: New file

	* getlyrics.pl: + changed to mach new Audio::MPD API

2008-06-01  Michal Nazarewicz  <mina86/AT/mina86.com>

	* arpping.c: New file

	* null.c: + -c switch added (for change directory)

2008-04-28  Michal Nazarewicz  <mina86/AT/mina86.com>

	* null.c: + fixed bug where -w was ignored when daemonising

2008-04-20  Michal Nazarewicz  <mina86/AT/mina86.com>

	* mpd-show.c: + fixed writing outside of buffer

2008-04-19  Michal Nazarewicz  <mina86/AT/mina86.com>

	* null.c: + -N switch for adjusting nice
	+ now waits for 1 sec in case child will terminate
	+ -w switch

2008-04-17  Michal Nazarewicz  <mina86/AT/mina86.com>

	* null.c: + removed configuration macros
	+ daemonisation also redirects stdin from /dev/null
	+ daemonisation also changes directory to root directory
	+ when discarding data from stdin changes directory to root directory
	+ nicer usage message added
	+ when checking command name only first character is analysed
	+ error checking added

2008-03-06  Michal Nazarewicz  <mina86/AT/mina86.com>

	* the-book-of-mozilla.txt: + two quotes added

2008-02-22  Michal Nazarewicz  <mina86/AT/mina86.com>

	* extractlinks.pl: + added -b option to specify base address

2008-02-07  Michal Nazarewicz  <mina86/AT/mina86.com>

	* FvwmTransFocus.c: + bug where prev_id value was not preserved fixed

2008-01-09  Michal Nazarewicz  <mina86/AT/mina86.com>

	* Makefile: + executes commands using 'exec' (less forks)

	* cdiff.c: + supports more formats
	+ may be run as a wrapper

	* cutcom.c: + new state as a pointer to new state, not index

	* FvwmTransFocus.c, check.sh, libmpdclient.c, libmpdclient.h,
	load.c, malloc.c, mountiso.c, mpd-show.c, mpd-state.c,
	one-time-pass.sh, timer.c, tuptime.c: + better standard compatibility
	+ better portability
	+ better usage of static and inline keywords (in case of C)

2007-08-21  Michal Nazarewicz  <mina86/AT/mina86.com>

	* getlyrics.pl: + added --edit option
	+ files in cache are saved in lower case

2007-08-10  Michal Nazarewicz  <mina86/AT/mina86.com>

	* Makefile: + mpd-show was missing in all target

------- New Release ----------------------------------------------------------


cdiff.c
/*
 * Colorize (adds ANSI codes) output of diff.
 * Copyright (c) 2005,2006 by Michal Nazareicz (mina86/AT/mina86.com)
 * Licensed under the Academic Free License version 3.0.
 *
 * This software is OSI Certified Open Source Software.
 * OSI Certified is a certification mark of the Open Source Initiative.
 *
 * This is part of Tiny Applications Collection
 *   -> http://tinyapps.sourceforge.net/
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>



/******************** Main ***************************************************/
static void not_tty(char **argv);
static int  run_diff(char **argv);
static void loop(void);

int main(int argc, char **argv) {
	/* If not TTY either run diff or pass in->out through (cat) */
	if (!isatty(1)) {
		not_tty(argv);
		return 1;
	}

	/* If arguments given run diff and prepare pipe */
	if (argc > 1) {
		int ret = run_diff(argv);
		if (ret) {
			return ret;
		}
	}

	/* Colourise output */
	loop();
	return 0;
}



/******************** Handle no TTY case *************************************/
static void not_tty(char **argv) {
	argv[0] = (char *)(argv[1] ? "diff" : "cat");
	execvp(argv[0], argv);
	perror(argv[1] ? "exec: diff" : "exec: cat");
}



/******************** Run diff and set up pipe *******************************/
static int run_diff(char **argv) {
	int fds[2];

	if (pipe(fds) < 0) {
		perror("pipe");
		return 1;
	}

	switch (fork()) {
	case -1:
		perror("fork");
		return 1;

	case 0:
		close(fds[0]);
		if (dup2(fds[1], 1) < 0) {
			perror("dup2");
			return 1;
		}
		argv[0] = (char *)"diff";
		execvp("diff", argv);
		perror("exec: diff");
		return 1;

	default:
		close(fds[1]);
		if (dup2(fds[0], 0) < 0) {
			perror("dup2");
			return 1;
		}
	}

	return 0;
}



/******************** Main loop; main part of the app ************************/
struct pattern {
	const char *pattern;
	unsigned idx;
};

static unsigned find(const struct pattern *patterns, const char *str);


static void loop(void) {
	enum color_number {
		COLOR_DEFAULT, /* must be zero */
		COLOR_INS,
		COLOR_DEL,
		COLOR_CHANGE,
		COLOR_EQUAL,
		COLOR_FILE1,
		COLOR_FILE2,
		COLOR_POS1,
		COLOR_POS2,
		COLOR_MISC
	};


	static const struct color {
		const char name[8];
		const char value[8];
	} colors[] = {
		{ "default", "0;37" },
		{ "ins" ,    "1;32" },
		{ "del" ,    "1;31" },
		{ "change",  "1;33" },
		{ "equal",   "0;37" },
		{ "file1",   "0;32" },
		{ "file2",   "0;31" },
		{ "pos1",    "0;32" },
		{ "pos2",    "0;31" },
		{ "misc",    "0;35" },
	};


	static const struct pattern unified_ruleset[] = {
		{ "+++ ", COLOR_FILE1   },
		{ "--- ", COLOR_FILE2   },
		{ "@@ " , COLOR_MISC    },
		{ "+"   , COLOR_INS     },
		{ "-"   , COLOR_DEL     },
		{ " "   , COLOR_EQUAL   },
		{ 0     , COLOR_DEFAULT },
	};

	static const struct pattern contex_ruleset[] = {
		{ "*** [^0-9]", COLOR_FILE1   },
		{ "--- [^0-9]", COLOR_FILE2   },
		{ "*** "      , COLOR_POS1    },
		{ "--- "      , COLOR_POS2    },
		{ "*"         , COLOR_MISC    },
		{ "?[^ ]"     , COLOR_MISC    },
		{ "+ "        , COLOR_INS     },
		{ "- "        , COLOR_DEL     },
		{ "! "        , COLOR_CHANGE  },
		{ "  "        , COLOR_EQUAL   },
		{ ""          , COLOR_DEFAULT },
	};

	static const struct pattern normal_ruleset[] = {
		{ "[0-9]", COLOR_MISC   },
		{ "---"  , COLOR_MISC   },
		{ "> "   , COLOR_INS    },
		{ "< "   , COLOR_DEL    },
		{ ""     , COLOR_DEFAULT },
	};

	static const struct pattern *rulesets[] = {
		0, /* must be at index zero */
		unified_ruleset,
		contex_ruleset,
		normal_ruleset
	};


	static const struct pattern modes[] = {
		{ "-",     1 },
		{ "*",     2 },
		{ "[0-9]", 3 },
		{ 0,       0 } /* side note: 0 == COLOR_DEFAULT */
	};


	const struct pattern *ruleset = 0;
	char buf[4096];


	while (fgets(buf, sizeof buf, stdin)) {
		unsigned idx;

		do {
			idx = find(ruleset ? ruleset : modes, buf);
		} while (!ruleset && (ruleset = rulesets[idx]));

		printf("\x1b[%sm%s", colors[idx].value, buf);
	}


	puts("\x1b[0m");
}



/******************** String matching ****************************************/
static int match(const char *pattern, const char *str);
static const char *match_group(const char *pat, char ch);


static unsigned find(const struct pattern *patterns, const char *str) {
	while (!match(patterns->pattern, str)) {
		++patterns;
	}
	return patterns->idx;
}


static int match(const char *pat, const char *str) {
	if (!pat) {
		return 1;
	}

	for (; *pat && *str; ++str, ++pat) {
		switch (*pat) {
		case '?':
			break;

		case '[':
			pat = match_group(pat, *str);
			if (!pat) return 0;
			break;

		default:
			if (*pat != *str) return 0;
		}
	}
	return 1;
}


static const char *match_group(const char *pat, char ch) {
	int ok = 0;

	++pat;
	if (*pat == '^') {
		ok = 1;
		++pat;
	}

	do {
		if (pat[1] != '-') {
			if (ch != *pat) continue;
		} else {
			const char from = *pat;
			const char to = *(pat += 2);

			if (from > to ? (to < ch   && ch < from)
			              : (ch < from || to < ch  )) continue;
		}

		/* Found */
		while (*++pat != ']') {
			/* nop */
		}
		ok = ok ^ 1;
		break;

	} while (*++pat != ']');

	return ok ? pat : 0;
}

